README.md
A Material Colors parsing library.

The colors_parser package provides string parsing operations for
MaterialColor and MaterialAccentColor values with an optional indexed
shade as well as black and white variants and even hex values.

## Installing ##

Use [pub][] to install this package. Add the following to your
`pubspec.yaml` file.

    dependencies:
      colors_parser: any

Then run `pub install`.

For more information, see the [colors_parser package on pub.dev][pkg].

[pub]: http://pub.dev
[pkg]: http://pub.dev/packages/colors_parser

## Using

The colors_parser package was designed to be imported with a prefix,
though you don't have to if you don't want to:

```dart
import 'package:colors_parser/colors_parser.dart' as cp;
```

The most common way to use the library is through the top-level functions.
These parse Material Colors based on English language names. For example:

```dart
cp.parse('blue');
```

This calls the top-level `parse()` function to parse "blue" using
English language names.

If you want to work with colors for a different language, you can create a
`ColorsParser` and give it an explicit colors map:

```dart
var cpES = new cp.ColorsParser(colorsMap: ["azul", Colors.blue]);
cpES.parse('azul');
```

This will parse "azul" using the supplied Spanish language colors map.


CHANGELOG.md
## 1.0.2

* Bump version for CHANGELOG.

## 1.0.0

* Initial release.


test/colors_parser_test.dart
// Copyright (c) 2020, Charles Esrock. All rights reserved.
// Use of this source code is governed by a BSD-style license
// that can be found in the LICENSE file.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:colors_parser/colors_parser.dart' as cp;

void main() {
  test('Parse colors primaries', () {
    for (var color in Colors.primaries) {
      testNamedColor(color);
    }
  });

  test('Parse colors accents', () {
    for (var color in Colors.accents) {
      testNamedColor(color);
    }
  });

  test('Parse special colors and black/white variants', () {
    testNamedColor(Colors.transparent);
    testNamedColor(Colors.grey);
    testNamedColor(Colors.black);
    testNamedColor(Colors.white);

    testNamedColor(Colors.black12);
    testNamedColor(Colors.black26);
    testNamedColor(Colors.black38);
    testNamedColor(Colors.black45);
    testNamedColor(Colors.black54);
    testNamedColor(Colors.black87);

    testNamedColor(Colors.white10);
    testNamedColor(Colors.white12);
    testNamedColor(Colors.white24);
    testNamedColor(Colors.white30);
    testNamedColor(Colors.white38);
    testNamedColor(Colors.white54);
    testNamedColor(Colors.white60);
    testNamedColor(Colors.white70);
  });

  test('Parse shades', () {
    for (var color in Colors.primaries) {
      for (var shade in [50, 100, 200, 300, 400, 500, 600, 700, 800, 900]) {
        testNamedColor(color[shade]);
      }
    }
    for (var color in Colors.accents) {
      for (var shade in [100, 200, 400, 700]) {
        testNamedColor(color[shade]);
      }
    }
  });

  test('Parse numeric colors', () {
    testNumericColor(Color(0x123456));
    testNumericColor(Color(0x12345678));
  });

  test('Parse exceptions', () {
    expect(cp.tryParse(null), isNull);
    expect(cp.tryParse(''), isNull);
    expect(cp.tryParse('[100]'), isNull);
    expect(cp.tryParse('unknown'), isNull);
    expect(cp.tryParse('blue10'), isNull);
    expect(cp.tryParse('blue[100'), isNull);
    expect(cp.tryParse('blue[]'), isNull);
    expect(cp.tryParse('blue[10]'), isNull);
    expect(cp.tryParse('black10'), isNull);
    expect(cp.tryParse('black12[100]'), isNull);
    expect(cp.tryParse('ff123456'), isNull);
    expect(cp.tryParse('0xff123456'), isNull);

    expect(cp.toColorsString(null), '');

    expect(cp.colorsMapEN['green'], Colors.green);
    expect(cp.colorsMap["blue"], Colors.blue);
  });
}

void testNamedColor(Color color) {
  var source = cp.toColorsString(color);
  expect(source, isNot(startsWith('#')));
  expect(cp.parse(source), color);
}

void testNumericColor(Color color) {
  expect(cp.parse(cp.toColorsString(color)), color);
}


example/lib/main.dart
// Copyright (c) 2020, Charles Esrock. All rights reserved.
// Use of this source code is governed by a BSD-style license
// that can be found in the LICENSE file.

import 'package:flutter/material.dart';

import 'package:colors_parser/colors_parser.dart' as cp;

void main() {
  runApp(App());
}

class App extends StatelessWidget {
  final colors = <String>[];

  App() {
    for (var color in Colors.primaries) colors.add(cp.toColorsString(color));
    for (var color in Colors.accents) colors.add(cp.toColorsString(color));
    for (var index in [50, 100, 200, 300, 400, 500, 600, 700, 800, 900])
      colors.add(cp.toColorsString(Colors.blue[index]));
    for (var index in [100, 200, 400, 700])
      colors.add(cp.toColorsString(Colors.blueAccent[index]));
    for (var color in ['grey', 'black', 'white']) colors.add(color);
    for (var color in [12, 26, 38, 45, 54, 87]) colors.add('black$color');
    for (var color in [10, 12, 30, 38, 54, 60, 70]) colors.add('white$color');
  }

  Widget build(BuildContext context) {
    var black = cp.parse('black');
    var grey500 = cp.parse('grey[500]');
    var white54 = cp.parse('white54');
    var black54 = cp.parse('black54');
    return MaterialApp(
      home: Scaffold(
          body: SafeArea(
              child: ListView.builder(
        itemCount: colors.length,
        itemBuilder: (BuildContext context, int index) {
          var color = cp.parse(colors[index]);
          var backColor = color.alpha == 0xFF ? color : grey500;
          var textColor =
              color.alpha != 0xFF ? color : color == black ? white54 : black54;
          return Container(
            height: 50,
            color: backColor,
            child: Center(
                child: Text(
                    '${colors[index]} #${cp.parse(colors[index]).value.toRadixString(16)}',
                    style: TextStyle(
                        fontSize: 20,
                        fontWeight: FontWeight.bold,
                        color: textColor))),
          );
        },
      ))),
    );
  }
}


lib/colors_parser.dart
// Copyright (c) 2020, Charles Esrock.  All rights reserved.
// Use of this source code is governed by a BSD-style license
// that can be found in the LICENSE file.

import 'package:flutter/material.dart';

import 'src/context.dart';

export 'src/context.dart';

/// The colors parser context.
///
/// This differs from a context created with [new Context] in that its
/// [Context.colorsMap] is always based on the English language, rather than
/// being set once when the context is created.
final Context context = Context();

/// Parses a string containing a Colors name or ARGB number into a Color
/// like [Context.parse].
Color parse(String source) => context.parse(source);

/// Parse [source] as a Colors name or ARGB number and return its value
/// like [Context.tryParse].
Color tryParse(String source) => context.tryParse(source);

/// Converts [color] to a Material Colors name or ARGB number
/// like [Context.toColorsString].
String toColorsString(Color color) => context.toColorsString(color);

/// Gets the colors map associated with the colors parser context
/// like [Context.colorsMap].
Map<String, Color> get colorsMap => context.colorsMap;


lib/src/context.dart
// Copyright (c) 2020, Charles Esrock. All rights reserved.
// Use of this source code is governed by a BSD-style license
// that can be found in the LICENSE file.

import 'package:flutter/material.dart';

/// An instantiable class for parsing colors. Unlike the top-level
/// functions, this lets you explicitly specify the colors map.
class Context {
  /// Creates a new colors parser context for the given colors map.
  ///
  /// If [colorsMap] is omitted, it uses the default English language
  /// colors map.
  Context({Map<String, Color> colorsMap})
      : colorsMap = colorsMap ?? colorsMapEN;

  /// The colors map that this context works with.
  final Map<String, Color> colorsMap;

  /// Parses a string containing a Colors name or ARGB number into a Color.
  ///
  /// The method first tries to read the [source] as a Colors MaterialColor
  /// or MaterialAccentColor name with an optional indexed shade. If that
  /// fails, it tries to read the [source] as a Colors black or white
  /// variant. If that also fails, it tries to parse the [source] as an
  /// ARGB integer. If that all fails, a [FormatException] is thrown.
  ///
  /// An ARGB integer [source] can include '#' for hex numbers. Make sure
  /// you specify 8 hex digits. If you only specify 6 hex digits, then the
  /// leading 2 digits default to 0, a fully transparent alpha value.
  ///
  /// Rather than throwing and immediately catching the [FormatException],
  /// instead use [tryParse] to handle a parsing error. Example:
  ///
  ///     var value = context.tryParse(text);
  ///     if (value == null) ... handle the problem
  Color parse(String source) {
    RegExpMatch match;

    match = RegExp(r'^([a-zA-Z]+([0-9]{0,3}))(\[([0-9]{2,3})\])?$')
        .firstMatch(source);
    if (match != null) {
      dynamic color = colorsMap[match.group(1)];
      if (color != null) {
        var index = match.group(4) == null ? 0 : int.parse(match.group(4));
        if (index == 0) return color;
        if (color[index] != null) return color[index];
      }
    }

    match = RegExp(r'^(\#?)([0-9a-fA-F]+)$').firstMatch(source);
    if (match != null) {
      var radix = match.group(1) != '' ? 16 : 10;
      var colorValue = int.tryParse(match.group(2), radix: radix);
      if (colorValue != null) return Color(colorValue);
    }

    throw FormatException('Invalid color name/number', source, 0);
  }

  /// Parse [source] as a Colors name or ARGB number and return its value.
  ///
  /// Like [parse] except that this method returns `null` where a similar call
  /// to [parse] would throw a [FormatException], and the [source] must still
  /// not be `null`.
  Color tryParse(String source) {
    try {
      return parse(source);
    } catch (e) {
      return null;
    }
  }

  /// Converts [color] to a Material Colors name or ARGB number.
  String toColorsString(Color color) {
    if (color != null) {
      for (var entry in colorsMap.entries) {
        if (entry.value == color) return entry.key;
        if (entry.value is MaterialColor) {
          MaterialColor materialColor = entry.value;
          for (var shade in [50, 100, 200, 300, 400, 500, 600, 700, 800, 900]) {
            if (materialColor[shade] == color) return '${entry.key}[$shade]';
          }
        }
        if (entry.value is MaterialAccentColor) {
          MaterialAccentColor materialAccentColor = entry.value;
          for (var shade in [100, 200, 400, 700]) {
            if (materialAccentColor[shade] == color)
              return '${entry.key}[$shade]';
          }
        }
      }
      return '#' + color.value.toRadixString(16);
    } else {
      return '';
    }
  }
}

/// Default English language colors map.
final colorsMapEN = {
  'red': Colors.red,
  'pink': Colors.pink,
  'purple': Colors.purple,
  'deepPurple': Colors.deepPurple,
  'indigo': Colors.indigo,
  'blue': Colors.blue,
  'lightBlue': Colors.lightBlue,
  'cyan': Colors.cyan,
  'teal': Colors.teal,
  'green': Colors.green,
  'lightGreen': Colors.lightGreen,
  'lime': Colors.lime,
  'yellow': Colors.yellow,
  'amber': Colors.amber,
  'orange': Colors.orange,
  'deepOrange': Colors.deepOrange,
  'brown': Colors.brown,
  'grey': Colors.grey,
  'blueGrey': Colors.blueGrey,
  'redAccent': Colors.redAccent,
  'pinkAccent': Colors.pinkAccent,
  'purpleAccent': Colors.purpleAccent,
  'deepPurpleAccent': Colors.deepPurpleAccent,
  'indigoAccent': Colors.indigoAccent,
  'blueAccent': Colors.blueAccent,
  'lightBlueAccent': Colors.lightBlueAccent,
  'cyanAccent': Colors.cyanAccent,
  'tealAccent': Colors.tealAccent,
  'greenAccent': Colors.greenAccent,
  'lightGreenAccent': Colors.lightGreenAccent,
  'limeAccent': Colors.limeAccent,
  'yellowAccent': Colors.yellowAccent,
  'amberAccent': Colors.amberAccent,
  'orangeAccent': Colors.orangeAccent,
  'deepOrangeAccent': Colors.deepOrangeAccent,
  'black': Colors.black,
  'black12': Colors.black12,
  'black26': Colors.black26,
  'black38': Colors.black38,
  'black45': Colors.black45,
  'black54': Colors.black54,
  'black87': Colors.black87,
  'white': Colors.white,
  'white10': Colors.white10,
  'white12': Colors.white12,
  'white24': Colors.white24,
  'white30': Colors.white30,
  'white38': Colors.white38,
  'white54': Colors.white54,
  'white60': Colors.white60,
  'white70': Colors.white70,
  'transparent': Colors.transparent
};
